kghDataManagement/views.py

```diff
@@ -1,3 +1,167 @@
+
+                        change[self.CHANGE_TITLE] = title
+                        change[self.CHANGE_KGH_ID] = newKghId
+
+                        kghItem.kghID = newKghId
+
+                        changes.append(change)
+
+                Item.objects.bulk_update(list(kghIdDict.values()), ["kghID", "price"])
+        except IntegrityError as e:
+            messages.error(request, f"Error: {str(e)}")
+            traceback.print_exc()
+            return redirect('kgh-upload')
+        except Exception as e:
+            messages.error(request, f"Error: {str(e)}")
+            traceback.print_exc()
+            return redirect('kgh-upload')
+
+        messages.success(request, "File upload successful")
+        context = {
+            "changes": changes,
+            "catalogUploaded": True
+        }
+        #  TODO, make a table of all changes (context already provided)
+        # If no changes are made, let the user know
+        # Use the global keys to reference each field
+
+        return render(request, 'kghDataManagement/kgh_upload_page.html', context=context)
+
+    # row: A row (as a dictionary) of the KGH catalog, as parsed from decode_utf8()
+    # kghItem: An [Item] that has a valid KGH ID (either outdated or present)
+    # Updates kghItem price and returns a dictionary of the changes made for the response context
+    def handleItemPriceChange(self, row, kghItem):
+        newPriceString = row.get(self.ROW_PRICE).strip()
+        newPrice = StringUtils.getFloatOrNone(newPriceString)
+        if newPrice == None:
+            raise Exception(
+                f"Could not parse material price for item with KGH ID: {kghItem.kghID}. Please make sure it is a decimal with no non-digit characters.")
+        oldPrice = kghItem.price
+        if float(oldPrice) != newPrice:
+            kghItem.price = newPrice
+            return ({
+                self.CHANGE_TITLE: kghItem.title,
+                self.CHANGE_OLD_PRICE: oldPrice,
+                self.CHANGE_NEW_PRICE: newPrice,
+                self.CHANGE_OLD_KGH_ID: kghItem.kghID
+            })
+        return {}
+
+    # oldIds: A space separated string of old KGH IDs
+    # itemDictionary: itemDictionary[kghID] == Item
+    def getItemMatchingOldIds(self, oldIds, itemDictionary):
+        ids = oldIds.split(" ")
+        for oldId in ids:
+            item = itemDictionary.get(oldId)
+            if item is not None:
+                return item
+
+    # Generator that decodes one row of the csv at a time
+    # Returns a value when iterated on, this way we don't need to store the entire decryption in memory
+    def decode_utf8(self, file):
+        # Skip the first row of the file which is the header
+        iterator = iter(file)
+        next(iterator)
+        for line in iterator:
+            row = line.decode('utf-8').replace('\r\n', '').split(',')
+            yield {
+                self.ROW_MATERIAL: row[0],
```

Comment:
I really don't like relying on column numbers, instead you should lookup column names in the header.

Reply:
Are we sure we want the csv to break if the catalog headers change slightly?
Also, please explain how you would do this. Are you saying store the first line in a variable and then do
self.ROW_MATERIAL: row[header.find("kgh ID")],

Reply:
Id trust the column names a lot more than they stay in an exact order without adding or removing columns. It also means you don't need dummy columns in the template. Your suggestion was what I had in mind